// zip-lib/lib/util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fs = require("fs");
const util = require("util");
function asyncCall(fn, ...args) {
    return new Promise((c, e) => fn(...args, (err, result) => err ? e(err) : c(result)));
}
function unlink(path) {
    if (typeof util.promisify === "function") {
        return util.promisify(fs.unlink)(path);
    }
    else {
        return asyncCall(fs.unlink, path);
    }
}
exports.unlink = unlink;
function mkdir(path, mode) {
    if (typeof util.promisify === "function") {
        return util.promisify(fs.mkdir)(path, mode);
    }
    else {
        return asyncCall(fs.mkdir, path, mode);
    }
}
exports.mkdir = mkdir;
function lstat(path) {
    if (typeof util.promisify === "function") {
        return util.promisify(fs.lstat)(path);
    }
    else {
        return asyncCall(fs.lstat, path);
    }
}
exports.lstat = lstat;
function chmod(path, mode) {
    if (typeof util.promisify === "function") {
        return util.promisify(fs.chmod)(path, mode);
    }
    else {
        return asyncCall(fs.chmod, path, mode);
    }
}
exports.chmod = chmod;
function readdir(path) {
    if (typeof util.promisify === "function") {
        return util.promisify(fs.readdir)(path);
    }
    else {
        return asyncCall(fs.readdir, path);
    }
}
exports.readdir = readdir;
function access(path, mode) {
    if (typeof util.promisify === "function") {
        return util.promisify(fs.access)(path, mode);
    }
    else {
        return asyncCall(fs.access, path, mode);
    }
}
exports.access = access;
function rmdir(path) {
    if (typeof util.promisify === "function") {
        return util.promisify(fs.rmdir)(path);
    }
    else {
        return asyncCall(fs.rmdir, path);
    }
}
exports.rmdir = rmdir;
function symlink(target, path) {
    if (typeof util.promisify === "function") {
        return util.promisify(fs.symlink)(target, path);
    }
    else {
        return asyncCall(fs.symlink, target, path);
    }
}
exports.symlink = symlink;
function readlink(path) {
    if (typeof util.promisify === "function") {
        return util.promisify(fs.readlink)(path);
    }
    else {
        return asyncCall(fs.readlink, path);
    }
}
exports.readlink = readlink;
